Feature/typescript

markogresak opened the pull request:
*Note: This is a work in progress and generator currently generates failing project.*

This pull request adds TypeScript option to generator, as asked in #106.

Current implementation:

 - [x] `--ts | --typescript` options
 - [x] generate working files (currently copied from js)
 - [ ] compile typescript files
 - [ ] ask if [tsd](https://github.com/DefinitelyTyped/tsd) should be added

Discuss:

 - [ ] Keep both `--ts | --typescript`, or just one of them?
 - [ ] Should compile `*.ts` > `*.js` or use [`tsconfig.json` project file](https://github.com/Microsoft/TypeScript/wiki/tsconfig.json)?

Also a question: When fixing/adding tests, I referred to coffee tests. There are some tests which tests working with coffee and ..., but when calling `runGenerationTest`, the `coffee` passed in is `false`. Why is that and should it be fixed? Example: [test-creation.js:159](https://github.com/petecoop/generator-express/blob/master/test/test-creation.js#L159).

petecoop commented:
- Having both --ts and --typescript sounds good to me. 
- I'm not sure on standards of compiling .ts files, but the basic idea of this project is that you can just run grunt or gulp and everything will watch/compile for you. Coffeescript does it through `require('coffee-script/register');`
- Yeah that looks like a problem with the tests, it'd probably be best to refactor the tests a bit so that you pass in a language var e.g. `coffee` or `typescript` instead of adding more true/false flags onto the function

markogresak commented:
I haven't tested it, but if I understand this correctly, coffee will compile the file and output it so same path, just replacing `.coffee` to `.js`, meaning that there are 2 files, source and the compiled javascript file in the same directory?
If I'm correct, I believe typescript with project config does that by default, but I'll have to disable my atom plugin and make sure it works correctly. I'd say it would be best to make it work with config file, because it's configurable by user and it's going to be de facto standard of TypeScript 1.5+. I know this works with gulp plugin, I'll have to check and confirm for gulp.

petecoop commented:
No the coffee files are compiled to js at runtime through that require, there's a similar thing [here](https://github.com/eknkc/typescript-require) for typescript

markogresak commented:
Oh okay, then I'll look into that a bit more. I've tested typescript-require, but it generates `.tmp` (or similar) folder and compiles files. `typescript-require` could be a quick solution, but form my experience, I see more and more devs use `tsconfig.json`. But with it, we would have to put up some rules or everything will break.
I'm for using `tsconfig.json` and setting it's `outDir` to a hidden folder, ignored by git and point the start script to that folder. It's a bit of a hack, but the end result isn't that much different compared to `typescript-require`. Editor plugins (I'm sure for atom-typescript) already require the file and will generate one for user if missing using file path as root. Created `tsconfig.json` is configured so files compile in the same directory, which all becomes just a huge mess.
So what do you think about using `tsconfig.json` with `outDir` set to `.ts-tmp` (or something similar)?

petecoop commented:
Makes sense if `tsconfig.json` is the standard, was just showing something similar to what's used in coffeescript.

Yeah your solution sounds good, maybe just call the outDir `build` or something to make it obvious?

markogresak commented:
I was just thinking of making hidden + ignored folder so it won't confuse users or so they won't to have set up it manually every time. Maybe not hidden, but ignored would be probably what most users would expect.
Or even better, ask them if it should be added to ignore folder.

petecoop commented:
I think `build` would make it more obvious and add it to the `.gitignore` but yes might be worth asking if it should be added to ignore, as some people would like to commit built files.

markogresak commented:
I have (finally) tried to finish this PR today since I had 3h of time. The commits might look a bit weird now, since I rebased some of commits I did into old ones, to avoid having too much pointless minifixes.

But the main problem is that I wasn't able to build typescript files yet.  I get error, which seems to be caused by sequelize:

```
~/generator-express-tests/node_modules/sequelize/lib/sequelize.js:626
      this.importCache[path] = defineCall(this, DataTypes);
                               ^
TypeError: defineCall is not a function
    at [object Object].module.exports.Sequelize.import (~/generator-express-tests/node_modules/sequelize/lib/sequelize.js:626:32)
    at ~/generator-express-tests/app/models/index.ts:8:36
    at Array.forEach (native)
    at Object.<anonymous> (~/generator-express-tests/app/models/index.ts:7:4)
    at Module._compile (module.js:430:26)
    at Object.loader (~/generator-express-tests/node_modules/typescript-node/dist/typescript-node.js:85:18)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Module.require (module.js:365:17)
    at require (module.js:384:17)
    at Object.<anonymous> (~/generator-express-tests/app.js:6:8)
    at Module._compile (module.js:430:26)
    at Object.Module._extensions..js (module.js:448:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Function.Module.runMain (module.js:471:10)
```

The `config.db` is set correctly before I use it. I do not have postgres server set up atm, but if that were the case, there would be a different error. I think it would be great to upgrade sequelize and then debug this error.


Also, I'm not sure how build is working with grunt and where is it watching for js/coffee files and how to add changes for building ts.
Although running `node app.js` should work, since I added `typescript-node/register` (I decided for this one, since it's the first choice of gulp when using gulpfile.ts).

petecoop commented:
It's looking good so far, but is it written in Typescript? Everything look the same but with a `.ts` extension

Have just updated the dependencies so might be worth a rebase and retrying with the latest.

To get the tests to run - you'll notice in the other functions they check if the value is already set, and if it is returns true e.g.:

```js
buildTool: function () {

      if (this.options.buildTool) {
        return true;
      }
}
```

This is just for the tests to skip the prompts. I think how the tests pass the options in needs a refactor really but will leave this till after this is done.

markogresak commented:
That's the point of TypeScript, it's just a superset of JavaScript. Anything that works in js should work in TypeScript, the only thing that sometimes needs fixing is specifying types. Once I get it working, I can refactor TypeScript to use more ES2015 like syntax.

Okay great, I will pull the changes and see if it helped. 

I expected that code to do the trick, but I just wasn't sure it's really doing what I expected so I asked instead of guessing and introducing bugs.

petecoop commented:
Ah right I see, yeah would be better if it had some Typescript features in there really.

markogresak commented:
Extra features come with typing files, but I haven't implemented tsd support yet. I have to first make it work with current code I'll be adding improvements later.

BTW I know these commits look terrible now, I will fix that before PR is accepted.

willsteinmetz commented:
I'm thinking given that this PR is about 3 years old, it might be worth hitting a reset button. Typescript is now at version 2.8.3 and a LOT has changed since 1.5. [ts-node](https://www.npmjs.com/package/ts-node) is the standard way to run Typescript code on Node. Also, DefinitelyTyped has been super fleshed out.

One of the comments above was that the code was mostly exactly the same with the file extension changed. While it's true that Typescript is a superset of JavaScript, it would probably be very worthwhile to developers using the package to go through and add in typings and convert code to the ES Module syntax to the template files.

That's just my $0.02 worth so take it or leave it as you wish :)

willsteinmetz commented:
Yeah, I don't even know if CoffeeScript is still being maintained. Last I read they weren't planning on making updates for ES2015 which was a non-starter for most developers. Pulling it out might make it easier to add in Typescript support.

I'm also wondering if it's worth keeping Ruby Sass support. Node Sass has come a long way and is a lot faster. It would also require fewer things to be installed to stick with just Node Sass.

mikeerickson commented:
Correct me if I am wrong, but I believe Ruby Sass is not going to be supported moving forward.  Further, the ruby build pipleline is now using Node Sass

willsteinmetz commented:
@mikeerickson I didn't know that but that's good to know. I swapped to Node Sass a few years ago. It just ran better and faster for me and didn't require me to install Ruby just for Sass.

I also looked on the CoffeeScript site. Looks like they reversed course and are supporting ES2015 transformations to some degree. Still not sure what the usage is now. I know when I did Rails a few years ago at one company we used it with that but I'm not sure if that's still the case because I didn't stick with Rails after that job because - and I know this is uncommon - I wasn't impressed.

mikeerickson commented:
@willsteinmetz Ya, I have never been on the CoffeeScript bandwagon myself, but I do know i that it has its followers (typically those who are in the Rails camp).   I personally use native ES2015 as it is pretty much 100% support on all modern browsers (just need to get module support)

willsteinmetz commented:
@mikeerickson I wonder if there are usage statistics anywhere for CoffeeScript? CoffeeScript was faster to write. For me, one of the big benefits of Typescript is that the typing helps save me from myself with regard to mistakes. I know it's just a preference, but it's one of the many reasons I stick with that over Coffeescript. If Typescript is not an option, like you I stick with ES2015+.

willsteinmetz commented:
@mikeerickson I opened #230 about the Sass thing mentioned above.

petecoop commented:
[CoffeeScript 2](http://coffeescript.org/announcing-coffeescript-2/) is where CoffeeScript is at the moment. I added a google analytics package to the generator to try and track installs earlier in the year to see if I can remove coffeescript - there's not much stats to draw from as I'm guessing it's not very reliable and people block it etc but based on what I've gathered it looks like CoffeeScript is used...
